futures-channel/src/mpsc/queue.rs

```diff
@@ -136,6 +137,32 @@ impl<T> Queue<T> {
 
         if self.head.load(Ordering::Acquire) == tail {Empty} else {Inconsistent}
     }
+
+    /// Pop an element similarly to `pop` function, but spin-wait on inconsistent
+    /// queue state instead of returning `Inconsistent`.
+    pub unsafe fn pop_spin(&self) -> Option<T> {
+        loop {
+            return match self.pop() {
+                Empty => None,
+                Data(t) => Some(t),
+                // Inconsistent means that there will be a message to pop
+                // in a short time. This branch can only be reached if
+                // values are being produced from another thread, so there
+                // are a few ways that we can deal with this:
+                //
+                // 1) Spin
+                // 2) thread::yield_now()
+                // 3) task::current().unwrap() & return Pending
+                //
+                // For now, thread::yield_now() is used, but it would
+                // probably be better to spin a few times then yield.
+                Inconsistent => {
+                    thread::yield_now();
+                    continue;
```

Comment:
uber nit: maybe `return` in the `Empty` and `Data` cases rather than `return` + `continue`?